sol/contracts/ics23.sol

```diff
@@ -0,0 +1,107 @@
+    }
+    function batchVerifyMembership(ProofSpec.Data memory spec, bytes memory commitmentRoot, CommitmentProof.Data memory proof, BatchItem[] memory items ) internal pure {
+        CommitmentProof.Data memory decoProof = Compress.decompress(proof);
+        for (uint i = 0; i < items.length; i++) {
+            verifyMembership(spec, commitmentRoot, decoProof, items[i].key, items[i].value);
+        }
+    }
+
+    function batchVerifyNonMembership(ProofSpec.Data memory spec, bytes memory commitmentRoot, CommitmentProof.Data memory proof, bytes[] memory keys ) internal pure {
+        CommitmentProof.Data memory decoProof = Compress.decompress(proof);
+        for (uint i = 0; i < keys.length; i++) {
+            verifyNonMembership(spec, commitmentRoot, decoProof, keys[i]);
+        }
+    }
+*/
+
+    // private
+    function getExistProofForKey(CommitmentProof.Data memory proof, bytes memory key) private pure returns(ExistenceProof.Data memory) {
+        if (ExistenceProof.isNil(proof.exist) == false){
+            if (BytesLib.equal(proof.exist.key, key) == true) {
+                return proof.exist;
+            }
+        } else if(BatchProof.isNil(proof.batch) == false) {
+            for (uint i = 0; i < proof.batch.entries.length; i++) {
+                if (ExistenceProof.isNil(proof.batch.entries[i].exist) == false && 
+                    BytesLib.equal(proof.batch.entries[i].exist.key, key)) {
+                    return proof.batch.entries[i].exist;
+                }
+            }
+        }
+        return ExistenceProof.nil();
+    }
+
+    function getNonExistProofForKey(CommitmentProof.Data memory proof, bytes memory key) private pure returns(NonExistenceProof.Data memory) {
+        if (NonExistenceProof.isNil(proof.nonexist) == false) {
+            if (isLeft(proof.nonexist.left, key) && isRight(proof.nonexist.right, key)) {
+                return proof.nonexist;
+            }
+        } else if (BatchProof.isNil(proof.batch) == false) {
+            for (uint i = 0; i < proof.batch.entries.length; i++) {
+                if (NonExistenceProof.isNil(proof.batch.entries[i].nonexist) == false && 
+                    isLeft(proof.batch.entries[i].nonexist.left, key) && 
+                    isRight(proof.batch.entries[i].nonexist.right, key)) {
+                    return proof.batch.entries[i].nonexist;
+                }
+            }
+        }
+        return NonExistenceProof.nil();
+    }
+
+    function isLeft(ExistenceProof.Data memory left, bytes memory key) private pure returns(bool) {
+        // ExistenceProof.isNil does not work
+        return ExistenceProof._empty(left) || Ops.compare(left.key, key) < 0;
+    }
+
+    function isRight(ExistenceProof.Data memory right, bytes memory key) private pure returns(bool) {
+        // ExistenceProof.isNil does not work
+        return ExistenceProof._empty(right) || Ops.compare(right.key, key) > 0;
+    }
+}
+
+
+contract ICS23_UnitTest {
```

Comment:
I don't know what is considered to be the best practice in solidity, but by looking at the [openzeppelin repo](https://github.com/OpenZeppelin/openzeppelin-contracts/tree/master/contracts/mocks) it seems that keeping the tests under the `mocks` repository is the thing.

I kinda like ^^ that, because the `contracts` and `libraries` are explicitly separated, wdyt?